### DISCUSSION TITLE
SPI bus sharing protocol

### DISCUSSION THREAD

**Discussion by mcbp223** [Bodmer/TFT_eSPI#1569] [New processor/display support or capability request] [answered]
Hi, I was using the protected data members _locked_ and _inTransaction_ in a derived class to follow the same protocol used by the library when accessing the same SPI bus, but now with commit [Add filled rectangles with gradient](https://github.com/Bodmer/TFT_eSPI/commit/0daee5d5020a0e49c139e73597d6c8ef90500c24#diff-c044b0e13c60e006f0323785077eb9b44936885da3984bc30144d67ed6d7af12) they've been made private. Any recommendations on how to guard access to the bus and play nicely with the library? I didn't get to test yet if anything bad happens if I don't check or set them.

And maybe a feature request, I'd like to use a 74HC138 line decoder to get 7 chip selects from 3 pins, or 4 chip selects from 2 pins. I was thinking maybe defining the CS_ macros outside the library would do? Like in platformio.ini.

---
Mainly using ESP32 here, but trying to keep projects compile for other targets too. Thanks for this comprehensive library btw, it's great!

**Comment by Bodmer** [comment, marked as answer]
The library supports transactions for multiple SPI users so access to those flags should not be necessary. The flags simply ensure that  SPI access requests and releases are always paired, otherwise the ESP32 will crash (though this may be fixed in the board package now).
The only time that you might have a problem is if you use tft.startWrite() which holds the chip select permanently low, in that case when tft.EndWrite is used then it waits for any ongoing SPI transaction to complete before releasing the chip select.

**Comment by mcbp223** [comment]
Thanks, I'll make sure I don't intersperse SPI calls without pairing them.

About the other request, I think I can add 4 additional SPI devices to the bus with just one more GPIO pin with a 74HC138, without any library modifications. For example if the 138 binary selection pins are A2, A1 and A0, and we use A0 as TFT_CS, A1 as TOUCH_CS and A2 for disambiguation, the truth table would be

```
A2   A1   A0   |  138_selection
---------------+---------------
0    0    0    |    0   invalid
0    0    1    |    1   touch
0    1    0    |    2   display
0    1    1    |    3   none
1    0    0    |    4   other_1
1    0    1    |    5   other_2
1    1    0    |    6   other_3
1    1    1    |    7   other_4
```
